patches/api/0322-Add-ItemStack-removeEnchantments.patch

```diff
@@ -0,0 +1,38 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: TehBrian <tehbrian@outlook.com>
+Date: Sat, 26 Jun 2021 14:58:50 -0400
+Subject: [PATCH] Add ItemStack#removeEnchantments
+
+
+diff --git a/src/main/java/org/bukkit/inventory/ItemStack.java b/src/main/java/org/bukkit/inventory/ItemStack.java
+index a7909406e9d54c1ab4789b984ed6b1da50837fce..135b6466ceb1f323bcf9a7703c6d2fc5894d3565 100644
+--- a/src/main/java/org/bukkit/inventory/ItemStack.java
++++ b/src/main/java/org/bukkit/inventory/ItemStack.java
+@@ -451,6 +451,27 @@ public class ItemStack implements Cloneable, ConfigurationSerializable, net.kyor
+         return level;
+     }
+ 
++    // Paper start - add #removeEnchantments
++    /**
++     * Removes the specified {@link Enchantment}s from this {@code ItemStack}.
++     * <p>
++     * This method is the same as calling {@link #removeEnchantment(Enchantment)}
++     * for each {@code Enchantment} in the given array.
++     *
++     * @param enchantments the {@code Enchantment}s to add
++     * @throws IllegalArgumentException if the given array of {@code Enchantment}s is {@code null}
++     *                                  or an element in the array is {@code null}
++     */
++    @Utility
++    public void removeEnchantments(@NotNull final Enchantment... enchantments) {
```

Comment:
Should this potentially return, similar to the single method, the levels of the enchantments removed ? Could just be an `int[]`. This is very much just a suggestion tho, I can definitely see how a utility method like this wouldn't need that feature.

Reply:
Ah, I had totally missed that `#removeEnchantment` had even returned anything in the first place! That's a great catch, thank you. The question is, should the method return an `int[]`, or should it return a `Map<Enchantment, Integer>`, which would be more consistent with the other enchantment methods? Additionally, if it returns a map, should enchantments which weren't removed have `0` as the value, or should they just not be included?

I can see a case made in favor of a simple int array, which is that you shouldn't need the enchantments that you pass in as arguments fed back to you via the return value. However, I think that the ease of use that a map provides would be beneficial. Not having to either think or care about the order which you pass in enchantments to be removed might reduce potential confusion. Furthermore, if we were to not include the enchantments which weren't actually removed, the return value must be a map by necessity, since the order of the enchantments fed in would not correspond to the indexes of the levels passed back.

Sorry about the length reply. Any thoughts?

Reply:
I'd definitely agree with your suggestions of a `Map<Enchantment, Integer>`. Regarding the resulting map, sticking to the spigot [removeEnchantment](https://papermc.io/javadocs/paper/1.17/org/bukkit/inventory/ItemStack.html#removeEnchantment(org.bukkit.enchantments.Enchantment)) style of returning level `0` for enchantment that did not exist might be a good approach, tho this is entirely subjective and more opinions would definitely help.

Reply:
Not including non-removed enchantments might improve usability, because instead of having to check the level, callers can simply check whether enchantments are present in the map to see if they were removed or not. I agree with what you said though, this is quite subjective. For now, just to get *something* out there, the current PR doesn't include non-removed enchantments in the map, so just let me know about any thoughts you have so far.